### PR TITLE
Runtime Dynamic Station Loading

### PR DESCRIPTION
**Pull request by Cyberboss**
TL;DR the main station map loaded along with SSmapping as opposed to when the world starts. This adds less than 15 seconds to SSmapping initializtion

Why?: Because compiling the map every time you want to change isn't kosher. Also this allows for map rotation without the server tools

Notable code changes:

- New map_config datums under SSmapping.config and SSmapping.next_map_config
- World default turf changed to turf/basic. It has a 0 overhead New() and ChangeTurf(). It's useful for speeding up world.maxz increases
- Map defines moved to __DEFINES
- The old defines (MAP_NAME, MAP_PATH, etc...) now exist in SSmapping.config
- maprotate and changemap moved to SSmapping
- Instead of those copypasta map .dm files we now have simple .json's
- See https://github.com/tgstation/tgstation/pull/24171/commits/4e4b481ed12b080be86fb4122cbaedb5198a4ed0 for new map specific job modifications

**TODO:**

- [x] Figure out a solution for custom .dm's (job changes, monastary shuttle)
- [x] Fix custom titlescreen icon_states (I know how, but I'm waiting on #24152)
- [x] Merge /datum/voteablemap into the new map_config datum
~~- [ ] Figure out how to runtime load centcomm without bugging out the splashscreen~~ Just gonna keep centcomm compiled
- [x] Kill off the old rotation tools
~~- [ ] Figure out how to handle server updates~~ See MSO's comment in #24552
- [x] Testmerges, testmerges, testmerges!

Closes #24549

@MrStonedOne, not sure who else to tag since I can't @tgstation/commit-access :/

:cl: Cyberboss
experiment: Map rotation has been made smoother
/:cl:

### CODE REVIEW

**Comment by coiax** [comment]
Yeah, the custom .dms are going to be a problem.

**Comment by RemieRichards** [comment]
>World default turf changed to turf/Basic. It has a 0 overhead New() and ChangeTurf(). It's useful for speeding up world.maxz increases

Maybe I can bring back my map instancing code...

**Comment by Quardbreak** [comment]
What about if i have custom lavaland-level and z-2? Where i can now make changes for custom picking z-levels?

**Comment by optimumtact** [comment]
I can't believe that this only adds 12 seconds to the startup, considering it's loading the maps at runtime in byond via the dmm suite

**Comment by Cyberboss** [comment]
I'll do better testing when I get home

**Comment by Jalleo** [comment]
The solution to custom dms with jobs could be either job pref seperated for all maps or datum creation at initialisation loading which jobs are meant to be there. In a sort of style lile our antag prefs are now.

**Comment by Cyberboss** [comment]
@optimumtact Real numbers (ruin generation disabled):

master:
Init: 8s
Mapping: 0.6s

This branch:
Init: <1s
Mapping: 13.2s

**Comment by optimumtact** [comment]
Hmm, impressive

**Comment by YotaXP** [comment]
> I can't believe that this only adds 12 seconds to the startup, considering it's loading the maps at runtime in byond via the dmm suite

Well I **did** improve the loader code last year. \*snaps suspenders\*

I never really liked the idea of map-specific DM files anyway.  I think it would be a good to come up with something better.

**Comment by Cyberboss** [comment]
Map specific stuff is taken care of: https://github.com/tgstation/tgstation/pull/24171/commits/4e4b481ed12b080be86fb4122cbaedb5198a4ed0

**Comment by Cyberboss** [comment]
Vacation till Sunday

**Comment by MrStonedOne** [comment]
@tgstation/commit-access 

i'll look at doing some test merges later if nobody else does.

**Comment by Cyberboss** [comment]
Seems free of obvious issues. Merge this and I'll revert 299d661 in a separate PR

**Comment by Cyberboss** [comment]
Need to figure out why uploaded map templates cant be placed

**Comment by Cyberboss** [comment]
Nevermind, that was a serparate issue

**Comment by Cyberboss** [comment]
Apparently, @MrStonedOne was having parallax problems if he joined before the map loaded, but I can't seem to reproduce it :/

**Comment by Cyberboss** [comment]
Strings aren't getting revaluated

**Comment by Incoming5643** [comment]
This pull breaks wizard teleport scrolls, which rely on a var named `teleportlocs`
The only place to teleport to is space (probably the initial 1,1,1 space tile that is z = 1)

**Comment by optimumtact** [comment]
@Cyberboss where are you at with this? Looking forward to getting it in.

**Comment by Cyberboss** [comment]
I guess you can go ahead and merge this. The only outstanding bug is dmm_suite not handling `\improper`. I'll fix that soonish.

File: code/__DEFINES/maps.dm [hunk 1]
```diff
@@ -27,5 +27,8 @@ Last space-z level = empty
 #define EMPTY_AREA_7 "Empty Area 7"
 #define EMPTY_AREA_8 "Empty Area 8"
 #define AWAY_MISSION "Away Mission"
-#define AWAY_MISSION_LIST list(AWAY_MISSION = SELFLOOPING)
-#define DEFAULT_MAP_TRANSITION_CONFIG list(MAIN_STATION = CROSSLINKED, CENTCOMM = SELFLOOPING, EMPTY_AREA_1 = CROSSLINKED, EMPTY_AREA_2 = CROSSLINKED, MINING = SELFLOOPING, EMPTY_AREA_3 = CROSSLINKED, EMPTY_AREA_4 = CROSSLINKED, EMPTY_AREA_5 = CROSSLINKED, EMPTY_AREA_6 = CROSSLINKED, EMPTY_AREA_7 = CROSSLINKED, EMPTY_AREA_8 = CROSSLINKED)
+
+//for modifying jobs
+#define MAP_JOB_CHECK if(SSmapping.config.map_name != JOB_MODIFICATION_MAP_NAME) { return; }
+#define MAP_JOB_CHECK_BASE if(SSmapping.config.map_name != JOB_MODIFICATION_MAP_NAME) { return ..(); }
+#define MAP_REMOVE_JOB(jobpath) /datum/job/jobpath/map_check() { return (SSmapping.config.map_name != JOB_MODIFICATION_MAP_NAME) && ..() }
```
Comment (RemieRichards) [inline code/__DEFINES/maps.dm]: should ```jobpath``` here not be ```##jobpath```? so it gets imported into the define correctly (# for import, ##for import no spaces)

File: code/modules/mob/new_player/new_player.dm [hunk 2]
```diff
@@ -17,6 +17,11 @@
 /mob/new_player/New()
 	tag = "mob_[next_mob_id++]"
 	mob_list += src
+	
+	if(client && !SSmapping.previous_map_config.defaulted && ticker.state == GAME_STATE_STARTUP)
+		testing("Found oldie")
```
Comment (RemieRichards) [inline code/modules/mob/new_player/new_player.dm]: butt

File: code/__DEFINES/maps.dm (hunk 1 above)
Comment (Cyberboss) [inline code/__DEFINES/maps.dm]: *shrug

jobpath is only every 1 word (e.g. `clown`, `hos`, `dectective`) and we don't have derivative jobs. Hasn't failed so far

Comment (RemieRichards) [inline code/__DEFINES/maps.dm]: See, this:
```MAP_REMOVE_JOB(hos) ... jobpath ...``` will become -> ```/datum/job/jobpath/map_check...``` See what's wrong? No "hos"

Ok, so if you imported it:
```MAP_REMOVE_JOB(hos) ... #jobpath ...``` -> ```/datum/job/ hos /mapcheck()```
What's wrong? Spaces in a path!

```MAP_REMOVE_JOB(hos) ... ##jobpath ...``` -> ```/datum/job/hos/mapcheck()```
Perfect!

**Comment by Cyberboss** [comment]
@MrStonedOne It just occurred to me that updating the server during a map rotation won't actually change the map or change it incorrectly because the json is in the opposite game folder. Am I correct?

**Comment by MrStonedOne** [comment]
what is the actual path of the json file?

**Comment by Cyberboss** [comment]
Fixed

File: _maps/loadallmaps.dm [hunk 3]
```diff
@@ -1,48 +1,14 @@
...
-#include "map_files\generic\z7.dmm"
-#include "map_files\generic\z8.dmm"
-#include "map_files\generic\z9.dmm"
-#include "map_files\generic\z10.dmm"
-#include "map_files\generic\z11.dmm"
-
-#undef TRAVIS_MASS_MAP_BUILD
-
-#ifdef TRAVISBUILDING
-#include "templates.dm"
```
Comment (MrStonedOne) [inline _maps/loadallmaps.dm]: What about the templates

Comment (Cyberboss) [inline _maps/loadallmaps.dm]: derp

File: code/__DEFINES/maps.dm (hunk 1 above)
Comment (YotaXP) [inline code/__DEFINES/maps.dm]: Have you tested this, Remie?  The single # definitely wouldn't work, as that encodes the argument as a string with double quotes, but I wouldn't think that the ## would be necessary either.  If it were, wouldn't you need it on both ends?

**Comment by MrStonedOne** [comment]
If you check the travis builds, you'll notice it isn't actually building any of the maps.

**Comment by Cyberboss** [comment]
Fixed

**Comment by MrStonedOne** [comment]
Now only one of them did so.

**Comment by Cyberboss** [comment]
The testing config isn't meant for building maps. It's just to make sure `#define TESTING` works

**Comment by MrStonedOne** [comment]
except the testing config is the only one that's building the maps.

**Comment by Iamgoofball** [comment]
So with this how close are we to no restart new rounds

**Comment by Cyberboss** [comment]
As soon as I genocide all globals

**Comment by RemieRichards** [comment]
Given byond's past, I'm *sure* skipping out on world reboots is going to cause memory leaks.

File: code/game/turfs/turf.dm [hunk 4]
```diff
@@ -52,6 +52,16 @@
 	CalculateAdjacentTurfs()
 
 /turf/Destroy(force)
+	. = QDEL_HINT_IWILLGC
+	if(force)
+		..()
+		//this will completely wipe turf state
+		var/turf/basic/B = new /turf/basic(src)
```
Comment (RemieRichards) [inline code/game/turfs/turf.dm]: won't this break a heck load of shit @Cyberboss @MrStonedOne ?
(atmos adjacent turfs as a guess)

Comment (Cyberboss) [inline code/game/turfs/turf.dm]: This is actually a premptive refactor for deleting a zlevel. Nothing actually uses it at the moment

**Comment by optimumtact** [comment]
14:30 <%Durkel> Servers being doing initialization for 3 minutes now
14:30 <%Durkel> am I being memed
14:30 <+cedarbridge> Its been super slow starting for a couple days now


Eeither this or your runtime initialization start changes have put a significant dent in our round start timer

**Comment by optimumtact** [comment]
14:33 <%Durkel> Last thing it loaded was the mining level, then it stopped.

File: code/game/turfs/basic.dm [hunk 5]
```diff
@@ -0,0 +1,6 @@
+//used to optimize map expansion
+/turf/basic/New()
+
+/turf/basic/ChangeTurf(var/T)
+    new T(src)
+    CRASH("Basic ChangeTurf")
```
Comment (optimumtact) [inline code/game/turfs/basic.dm]: ?

Comment (Cyberboss) [inline code/game/turfs/basic.dm]: This shouldnt be being called!

Comment (optimumtact) [inline code/game/turfs/basic.dm]: hmm well it was called 500,000 times!

File: code/controllers/subsystem/mapping.dm [hunk 6]
```diff
@@ -81,6 +96,101 @@ var/datum/subsystem/mapping/SSmapping
...
+
+	INIT_ANNOUNCE("Loading [config.map_name]...")
+	TryLoadZ(config.GetFullMapPath(), FailedZs, 1)
+	INIT_ANNOUNCE("Loaded station!")
+
+	INIT_ANNOUNCE("Loading mining level...")
+	TryLoadZ("_maps/map_files/generic/[config.minetype].dmm", FailedZs)
+	INIT_ANNOUNCE("Loaded mining level!")
+
+	for(var/I in 1 to 5)
```
Comment (Someguynamedpie) [inline code/controllers/subsystem/mapping.dm]: for one, use `world.maxz` and `world.turf` for two `for(var/i in ZLEVEL_SPACEMIN to ZLEVEL_SPACEMAX)` is just a few lines above use the constants

File: code/controllers/subsystem/mapping.dm [hunk 7]
```diff
@@ -81,6 +96,101 @@ var/datum/subsystem/mapping/SSmapping
...
+
+	//filter votes
+	for (var/map in mapvotes)
+		if (!map)
+			mapvotes.Remove(map)
+		if (!(map in global.config.maplist))
+			mapvotes.Remove(map)
+			continue
+		var/datum/map_config/VM = global.config.maplist[map]
+		if (!VM)
```
Comment (Cyberboss) [inline code/controllers/subsystem/mapping.dm]: It's just ported from the old system

**Comment by Cyberboss** [comment]
Ready af

**Comment by YotaXP** [comment]
> 97 Commits

oh bby :tada: 
Can't merge this until we get 3 more.

**Comment by Cyberboss** [comment]
KOOOOOOOOOOOOOOOOOORRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRR!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by Cyberboss** [comment]
Had to tear out titlescreen transition smoothing DNM yet 😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠😠

**Comment by lzimann** [comment]
Two more commits let's go !

**Comment by optimumtact** [comment]
I'll merge it man I aint playing

**Comment by Cyberboss** [comment]
DO IT BEFORE I LOSE IT @optimumtact

**Comment by Cyberboss** [comment]
Yeehaw

**Comment by PKPenguin321** [comment]
exactly 100 commits NICE

**Comment by xxalpha** [comment]
REVERT THIS
